core/capabilities/targets/write_target.go

```diff
@@ -0,0 +1,232 @@
+	chain legacyevm.Chain
+	capabilities.CapabilityInfo
+}
+
+func NewEvmWrite(chain legacyevm.Chain) *EvmWrite {
+	// TODO: generate ID based on network id + chain id
+	return &EvmWrite{
+		chain,
+		info,
+	}
+}
+
+type EvmConfig struct {
+	ChainID uint
+	Address string
+	Params  []any
+	ABI     string
+}
+
+// TODO: enforce required key presence
+
+func parseConfig(rawConfig *values.Map) (EvmConfig, error) {
+	var config EvmConfig
+	configAny, err := rawConfig.Unwrap()
+	if err != nil {
+		return config, err
+	}
+	err = mapstructure.Decode(configAny, &config)
+	return config, err
+}
+
+func evaluateParams(params []any, inputs map[string]any) ([]any, error) {
+	vars := pipeline.NewVarsFrom(inputs)
+	var args []any
+	for _, param := range params {
+		switch v := param.(type) {
+		case string:
+			get := pipeline.VarExpr(v, vars)
+			val, err := get()
+			if err == nil {
+				args = append(args, val)
+			} else if errors.Is(errors.Cause(err), pipeline.ErrParameterEmpty) {
+				args = append(args, param)
+			} else {
+				return args, err
+			}
+		default:
+			args = append(args, param)
+		}
+	}
+
+	return args, nil
+}
+
+func encodePayload(args []any, rawSelector string) ([]byte, error) {
+	// TODO: do spec parsing as part of parseConfig()
+
+	// Based on https://github.com/ethereum/go-ethereum/blob/f1c27c286ea2d0e110a507e5749e92d0a6144f08/signer/fourbyte/abi.go#L77-L102
+
+	// NOTE: without having full ABI it's actually impossible to support function overloading
+	selector, err := abiutil.ParseSignature(rawSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	abidata, err := json.Marshal([]abi.SelectorMarshaling{selector})
+	if err != nil {
+		return nil, err
+	}
+
+	spec, err := abi.JSON(strings.NewReader(string(abidata)))
+	if err != nil {
+		return nil, err
+	}
+
+	return spec.Pack(selector.Name, args...)
+
+	// NOTE: could avoid JSON encoding/decoding the selector
+	// var args abi.Arguments
+	// for _, arg := range selector.Inputs {
+	// 	ty, err := abi.NewType(arg.Type, arg.InternalType, arg.Components)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// 	args = append(args, abi.Argument{Name: arg.Name, Type: ty})
+	// }
+	// // we only care about the name + inputs so we can compute the method ID
+	// method := abi.NewMethod(selector.Name, selector.Name, abi.Function, "nonpayable", false, false, args, nil)
+	//
+	// https://github.com/ethereum/go-ethereum/blob/f1c27c286ea2d0e110a507e5749e92d0a6144f08/accounts/abi/abi.go#L77-L82
+	// arguments, err := method.Inputs.Pack(args...)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// // Pack up the method ID too if not a constructor and return
+	// return append(method.ID, arguments...), nil
+}
+
+func (cap *EvmWrite) Execute(ctx context.Context, callback chan<- capabilities.CapabilityResponse, request capabilities.CapabilityRequest) error {
+	// TODO: technically we need to dispatch to the correct chain via network+chainID, so cap needs to store chainsets rather than a single chain
```

Comment:
I don't think that's true. I think we'll create separate capability instances per chain.